### PR TITLE
+keydown controls; -mousemove controls; +pointerlock

### PR DESCRIPTION
**Pull request by Droxus**

### CODE REVIEW

File: src/core/Controls.ts [hunk 1]
```diff
@@ -1,24 +1,75 @@
...
         const { width, height } = context.viewport;
-        this.camera.aspect = width / height;
-        this.camera.updateProjectionMatrix();
+        Controls.camera.aspect = width / height;
+        Controls.camera.updateProjectionMatrix();
+    }
+    public static createControls(): void {
+        let speed = 10, sensitivity = 1
+        // flyControls for comfortable devloping   
+        let canvas = document.querySelector('#canvas') as HTMLCanvasElement;
```
Comment (sergioness) [inline src/core/Controls.ts]: use `View.getDomElement()`

File: src/core/Controls.ts [hunk 2]
```diff
@@ -1,24 +1,75 @@
...
+import { Game } from "./Game";
 /**
  * An instance of this class perceives a user's actions within the game,
  * and controls the game input.
  */
 export class Controls {
-    private camera: THREE.PerspectiveCamera;
+    private static camera: THREE.PerspectiveCamera;
+    public static controls: any;
+    public static euler = new THREE.Euler( 0, 0, 0, 'YXZ' );
```
Comment (sergioness) [inline src/core/Controls.ts:11]: why do we need `static` fields here?

File: src/core/Controls.ts [hunk 3]
```diff
@@ -1,24 +1,75 @@
...
+    }
+    public static createControls(): void {
+        let speed = 10, sensitivity = 1
+        // flyControls for comfortable devloping   
+        let canvas = document.querySelector('#canvas') as HTMLCanvasElement;
+          const canvasClick = () => {
+            canvas.requestPointerLock();
+          };
+          
+          document.addEventListener('pointerlockchange', () => {console.log('sam loh')}, false);
```
Comment (sergioness) [inline src/core/Controls.ts]: remove the offensive comment

File: src/core/Controls.ts [hunk 4]
```diff
@@ -1,24 +1,75 @@
...
-        this.camera.aspect = width / height;
-        this.camera.updateProjectionMatrix();
+        Controls.camera.aspect = width / height;
+        Controls.camera.updateProjectionMatrix();
+    }
+    public static createControls(): void {
+        let speed = 10, sensitivity = 1
+        // flyControls for comfortable devloping   
+        let canvas = document.querySelector('#canvas') as HTMLCanvasElement;
+          const canvasClick = () => {
```
Comment (sergioness) [inline src/core/Controls.ts]: create a respective method on this class, and pass its reference to the `<model>.addEventListener()`'s 2nd argument with help of `.bind(this)`, e.g.
```typescript
private _onDomElementClick(): void {
    this.domElement.requestPointerLock();
}
private _createControls(): void {
    this.domElement.addEventListener('click', this._onDomElementClick.bind(this), false);
}
```

File: src/core/Controls.ts [hunk 5]
```diff
@@ -1,24 +1,75 @@
...
+            canvas.requestPointerLock();
+          };
+          
+          document.addEventListener('pointerlockchange', () => {console.log('sam loh')}, false);
+          canvas.addEventListener('click', canvasClick, false);
+          
+          canvas.addEventListener("mousemove", (event: any) => {
+            console.log('asd')
```
Comment (sergioness) [inline src/core/Controls.ts]: remove unnecessary logs

File: src/core/Controls.ts [hunk 6]
```diff
@@ -1,24 +1,75 @@
...
+            const movementX = event.movementX || event.mozMovementX || event.webkitMovementX || 0;
+            const movementY = event.movementY || event.mozMovementY || event.webkitMovementY || 0;
+        
+            Controls.euler.x -= movementY / (1 / sensitivity * 1000)
```
Comment (sergioness) [inline src/core/Controls.ts]: where does the `1000` magic number come from?

File: src/core/Controls.ts [hunk 7]
```diff
@@ -1,24 +1,75 @@
...
+            Controls. euler.y -= movementX / (1 / sensitivity * 1000) 
+        
+            Controls.euler.x = Math.max(Math.min(Math.PI/2, Controls.euler.x), -Math.PI/2)
+            Controls.camera.quaternion.setFromEuler( Controls.euler );
+            Game.needRender = true
```
Comment (sergioness) [inline src/core/Controls.ts]: how about introducing a new method on this class, say, `addListener(callback)` which would take a listener callback which would be triggered it everytime these controls would be changed?

File: src/core/Controls.ts [hunk 8]
```diff
@@ -1,24 +1,75 @@
...
+        });
+        window.addEventListener('keydown', (event) => {
+            console.log(event.code)
+            switch (event.code) {
```
Comment (sergioness) [inline src/core/Controls.ts]: [nit]
prefer a map `keydowns` with `event.code`s as its keys and the respective `switch`-`case` blocks as its values.

For instance,
```typescript
    private cameraSpeed = 10;
    private keydowns = new Map([
        ['KeyW', () => this.translateCameraTo({ z: -this.cameraSpeed})],
        ['KeyD', () => this.translateCameraTo({ x: this.cameraSpeed})],
        // ...
    ]);
    private translateCameraTo(params: Partial<{ x: number, y: number, z: number }>): void {
        const { x, y, z } = params;
        if (x) this.camera.translateX(x);
        if (y) this.camera.translateY(y);
        // ...
    }
    private _onKeyDown(event: KeyboardEvent): void {
        if (this.keydowns.has(event.code)){
            const processThisEventCode = this.keydowns.get(event.code);
            if (processThisEventCode) processThisEventCode();
        }
        this._fireChangeEvent();
    }
    private _fireChangeEvent(): void {
        // call the callback received in the addListener(callback) method of this class,
        // which would make the Game class's instance to re-render its scene
    }
    public addListener(callback: (...args: any) => any): void {
        // keep reference to the callback function, and call it when a change occurs
    }
```

File: src/core/Game.ts [hunk 9]
```diff
@@ -21,10 +21,10 @@ export class Game {
     /** {@link requestAnimationFrame} request ID to use with {@link cancelAnimationFrame} */
     private animationRequestId: any;
     /** tells the game loop whether its scene needs to be re-rendered (true) or not */
-    private needRender: boolean;
+    public static needRender: boolean;
```
Comment (sergioness) [inline src/core/Game.ts]: why do we mark `needRender` as `public static`?
there should be another way of accessing private properties from an instance of one class in another

File: src/core/Controls.ts [hunk 10]
```diff
@@ -1,24 +1,75 @@
...
      */
     public update(context: { viewport: { width: number, height: number } }): void {
         const { width, height } = context.viewport;
-        this.camera.aspect = width / height;
-        this.camera.updateProjectionMatrix();
+        Controls.camera.aspect = width / height;
+        Controls.camera.updateProjectionMatrix();
+    }
+    public static createControls(): void {
+        let speed = 10, sensitivity = 1
```
Comment (sergioness) [inline src/core/Controls.ts]: consider making these properties as this class citizens

**Review by sergioness — CHANGES_REQUESTED** [review]
I wish Three.js controls would be this efficient!
Great work! Keep it up!

**Comment by sergioness** [comment]
> I wish Three.js controls would be this efficient! Great work! Keep it up!

please, see my PR #10 to address these comments accordingly

File: src/core/Controls.ts (hunk 6 above)
Comment (Droxus) [reply src/core/Controls.ts]: Try to remove this 'magic' number and let me see your aim. This need to reduce input sensetivity to 'comfortable' value

File: src/core/Controls.ts (hunk 3 above)
Comment (Droxus) [reply src/core/Controls.ts]: don't worry, this is not about you =D

File: src/core/Controls.ts (hunk 5 above)
Comment (Droxus) [reply src/core/Controls.ts]: will be done

File: src/core/Controls.ts (hunk 2 above)
Comment (Droxus) [reply src/core/Controls.ts:11]: Because for errors without this, but i will check how to avoid it

File: src/core/Controls.ts (hunk 8 above)
Comment (Droxus) [reply src/core/Controls.ts]: Nice refactoring, will be done

File: src/core/Controls.ts (hunk 1 above)
Comment (Droxus) [reply src/core/Controls.ts]: Can not introduce it, need your additional explaining

File: src/core/Controls.ts (hunk 10 above)
Comment (Droxus) [reply src/core/Controls.ts]: will be done

**Pull request merged**